File: dynamic_programming_1/day_1/1137_NthTribonacci/Tribonacci.py
## BASE Code passing testcases
class Solution:
    def tribonacci(self, n: int) -> int:
        t0, t1, t2 = 0, 1, 1
        if n == 0:
            return t0
        if n == 1:
            return t1
        if n == 2:
            return t2
        for i in range(3,n+1):
            sum = t0 + t1 + t2
            t0, t1, t2 = t1, t2, sum
        return sum

## Solution using Arrays
class Solution:
    def tribonacci(self, n: int) -> int:
        t = [0, 1, 1]
        if n == 0:
            return t[0]
        if n == 1:
            return t[1]
        if n == 2:
            return t[2]
        for i in range(3,n+1):
            t.append(t[i-1] + t[i-2] + t[i-3])
        return t[n]

## More conscise solution 
class Solution:
    def tribonacci(self, n: int) -> int:
        if(n==0):
            return 0
        a, b, c = 0, 1, 1
        for _ in range(n-2): a, b, c = b, c, a+b+c
        return c

File: dynamic_programming_1/day_1/1137_NthTribonacci/Tribonacci.cpp
// BASE code that exceeded Time Limit
class Solution {
public:
    int tribonacci(int n) 
    {
        if(n == 0) {return 0;}
        if(n == 1) {return 1;}
        if(n == 2) {return 1;}
        return(tribonacci(n-1) + tribonacci(n-2) + tribonacci(n-3)) 
    }
};

// Fastest Code using arrays passed testcases
class Solution {
public:
    int tribonacci(int n) 
    {
        if(n == 0) {return 0;}
        if(n == 1) {return 1;}
        if(n == 2) {return 1;}
        vector<int> trib(n+1);
        trib[0] = 0;
        trib[1] = 1;
        trib[2] = 1;
        for(int i = 3; i <= n; i++)
        {
            trib[i] = trib[i-1]+trib[i-2]+trib[i-3];
        }
        return trib[n];
    }
};

// Fastest code with constant memory usage
class Solution {
public:
    int tribonacci(int n) 
    {
        if(n == 0) {return 0;}
        if(n == 1) {return 1;}
        if(n == 2) {return 1;}
        uint first = 0, second = 1, third = 1, result;
        while((n--)-2)
        {
            result = first + second + third;
            first = second;
            second = third;
            third = result;
        }
        return result;
    }
};